### ISSUE TITLE
Should QA tests be able to test with assertions enabled?

### ISSUE THREAD

**Issue by samreid** [phetsims/tasks#722]
During https://github.com/phetsims/phet-io/issues/789 I realized that when we deliver RCs or Dev versions to the QA team, they cannot test assertions.  Generally speaking, there is a class of problems where knowing if an assertion is triggered would be valuable, which is not being tested in RC or Dev testing.  Should we provide a way to make this happen?  Perhaps a new rule that "dev" tests have assertions enabled, and "rc" tests do not?  Or is the weekly automated test sufficient?  This issue is for discussion at next developer meeting.

**Comment by samreid** [comment]
@ariel-phet: our QA team is thorough enough that we do not need more extensive testing.  We may want to make sure @phet-steele and the other QA team know how to use phet-test.colorado.edu to run a requirejs sim with assertions enabled to get further information about bugs found in published versions.

**Comment by samreid** [comment]
@ariel-phet says it may be good to discuss with @phet-steele, perhaps it may be good in early dev testing.

**Comment by samreid** [comment]
@phet-steele what is your opinion about testing with assertions?  Do you ever use this strategy on phettest.colorado.edu (by launching with ?ea).

**Comment by phet-steele** [comment]
> Do you ever use this strategy on phettest.colorado.edu (by launching with ?ea)?

Extensively, all the time, abundantly.


> what is your opinion about testing with assertions?

Assertions are great, but only useful when not minified. I take this to be the reason why assertions are filtered out during the build process.

My current process:
1. Receive a dev or rc version to test.
2. Using two monitors, one has me directly interacting with the dev or rc
3. The other passively runs with ?fuzzMouse, but is not running the dev or rc. It's running the most up-to-date master on phettest with ?ea.

This is risky, since the code being tested on phettest might be different than the code in the dev or rc, or even unstable. If I ever hit an assert on phettest, I take it with a grain of salt. I try my best to replicate the error by hand on phettest (which is usually successful) and ensure that buggy behaviour exists in the dev/rc with the same procedure. From there, I would report the phettest assertion alongside a procedure, stating a caveat along the lines of "this is from phettest, but also happens in the dev/rc".

This whole process is clunky and has me spending a fair amount of time not testing the dev/rc code, but I live with it. I've been okay with this because minified assertions are much more useless than non minified ones.


> Perhaps a new rule that "dev" tests have assertions enabled, and "rc" tests do not?

The only thing that this would solve (whether it be dev only or both dev and rc) is that my step 3 process would be guaranteed to be testing the correct, stable code. Would this save me a bunch of time? You bet. Would I be getting useful errors that make a developer's life easier? Nope. Fact of the matter is, I have plenty of time to test. Developers have much tighter schedules than I. We don't need to change anything to suit my needs.

With regards to the rest of the team, they know the tools at their disposal. Although, they do not have the same amount of time in their day to follow my process. This means they usually stick with testing the dev/rc and avoid phettest. This is a better use of their time. So for their sake? Yes, assertions would be nice to have in a dev/rc so they can encounter problems they would have otherwise never seen.

In summary, keep the assertions in a dev version. This would make student tester time more valuable. This also would not impede me, but would only serve to verify my findings from phettest. I probably seemed back and forth on this topic, but there were some pros and cons to point out.

**Comment by samreid** [comment]
@phet-steele recommended that we keep assertions in dev versions.  Perhaps we should also publish dev versions without minification, so the assertion errors will have a more meaningful stack trace.  Assigned for developer meeting to decide how to proceed.

**Comment by jonathanolson** [comment]
FYI, building with ```--uglify=false``` and running with ```?ea``` doesn't currently run assertions.

**Comment by samreid** [comment]
> FYI, building with --uglify=false and running with ?ea doesn't currently run assertions.

So do we need some other flag to avoid stripping assertions?

**Comment by jonathanolson** [comment]
```js
var enableAllAssertions = !isProduction && phet.chipper.queryParameters.eall; // enables all assertions (basic and slow)
    var enableBasicAssertions = enableAllAssertions || ( !isProduction && phet.chipper.queryParameters.ea );  // enables basic assertions
```
it is not stripped, just explicitly disabled.

**Comment by samreid** [comment]
@pixelzoom suggests seeing how difficult it would be to create both forms during the build.

**Comment by samreid** [comment]
@jonathanolson suggests making it easy for @phet-steele to checkout-shas from a published dev version.

**Comment by samreid** [comment]
Perhaps I can get to this in a month or so.  But @ariel-phet said @jbphet or @jessegreenberg may be able to do this.

**Comment by samreid** [comment]
@jonathanolson I like your suggestion in https://github.com/phetsims/tasks/issues/722#issuecomment-264280233, were you thinking that @phet-steele would do this on phettest or some server?  Or on a development machine?

**Comment by jonathanolson** [comment]
> @jonathanolson I like your suggestion in #722 (comment), were you thinking that @phet-steele would do this on phettest or some server? Or on a development machine?

Presumably not phettest, since other people use it to test master. Probably would have things set up on his primary computer?

**Comment by phet-steele** [comment]
> Probably would have things set up on his primary computer?

The idea was to have assertions available for student testers, on all of our test devices, not just me on one device. Might not be the most efficient to set up whatever you had in mind on ~25 different devices? (unless I'm misunderstanding)

**Comment by samreid** [comment]
@jonathanolson what about another checkout on phettest where @phet-steele can paste in a dependencies.json into a text box, and it will pull+check out those branches?  The main phettest would remain on master.

**Comment by jonathanolson** [comment]
Due to the need to have multiple testers use this (and presumably not coordinate with each other), it sounds like we should probably include or change deployments to have a version with assertions available.

I'll investigate adding this ability to the build (at least for builds that get deployed, preferably having both forms built)

**Comment by jonathanolson** [comment]
> I forget what we decided and what's been implemented.

The --debugHTML flag triggers the building of the _debug version. Deployments for perennial pretty much all build this when possible. For phet-brand, it is not uglified (phet-io minifies and mangles without stripping assertions). Assertions are turned on for these builds (the `phet.chipper.isDebugBuild` flag is set to true in the template).

**Comment by samreid** [comment]
Sounds great @jonathanolson, reassigning to @phet-steele to evaluate if/how QA should leverage these builds, if at all.

**Comment by phet-steele** [comment]
QA knows the debug versions exist, and they know when to use them 👍
